lua/ulx/modules/urs_server.lua

```diff
@@ -1,193 +1,268 @@
-		if URS.limits[type][ply:SteamID()] then 
-			if ply:GetCount(type.."s") >= URS.limits[type][ply:SteamID()] then 
-				ply:LimitHit( type .."s" )
-				return false 
-			end 
-		elseif URS.limits[type][group] then 
-			if ply:GetCount(type.."s") >= URS.limits[type][group] then 
-				ply:LimitHit( type .."s" )
-				return false 
-			end 
-		end 
-		if URS.cfg.overwriteSbox:GetBool() then 
-			return true -- Overwrite sbox limit (ours is greater) 
-		end 
-	end 
+    rawset( restrictionTypes, what, result )
+
+    return result
 end
+local cacheCheck = URS.CacheCheck
+
+function URS.Check(ply, restrictionType, what)
+    local cachedResult = cachedCheck( ply, restrictionType, what )
+    if cachedResult ~= nil then return cachedResult end
+
+    what = stringLower(what)
+    local restrictionTypePlural = restrictionType .. "s"
+    local group = ply:GetUserGroup()
+
+    local restriction = false
+
+    local ursRestrictions = rawget( URS, "restrictions" )
+    local typeRestrictions = ursRestrictions and rawget( ursRestrictions, restrictionType )
+    local theseRestrictions = typeRestrictions and rawget( typeRestrictions, what )
 
-timer.Simple(0.1, function() 
-
-	--  Wiremod's Advanced Duplicator
-	if AdvDupe then 
-		AdvDupe.AdminSettings.AddEntCheckHook( "URSDupeCheck", 
-		function(ply, Ent, EntTable) 
-			return URS.Check( ply, "advdupe", EntTable.Class )
-		end, 
-		function(Hook) 
-			ULib.tsayColor( nil, false, Color( 255, 0, 0 ), "URSDupeCheck has failed.  Please contact Aaron113 @\nhttp://forums.ulyssesmod.net/index.php/topic,5269.0.html" )
-		end )
-	end
-
-	-- Advanced Duplicator 2 (http://facepunch.com/showthread.php?t=1136597)
-	if AdvDupe2 then 
-		hook.Add("PlayerSpawnEntity", "URSCheckRestrictedEntity", function(ply, EntTable) 
-			if URS.Check(ply, "advdupe", EntTable.Class) == false or URS.Check(ply, "advdupe", EntTable.Model) == false then 
-				return false 
-			end 
-		end) 
-	end 
+    if theseRestrictions then
+        restriction = theseRestrictions
+    end
+
+    if restriction then
+        local hasGroup = rawget( restriction, group )
+        local hasPlayer = rawget( restriction, ply:SteamID() )
+
+        if rawget( restriction, "*" ) then
+            if not ( hasGroup or hasPlayer ) then
+                PrintRestricted( ply, restrictionType, what )
+                return cacheCheck( ply, restrictionType, what, false )
+            end
+        elseif hasGroup or hasPlayer then
+            PrintRestricted( ply, restrictionType, what )
+            return cacheCheck( ply, restrictionType, what, false )
+        end
+    end
+
+    local allRestrictions = rawget( ursRestrictions, "all" )
+    local allTypeRestrictions = allRestrictions and rawget( allRestrictions, restrictionType )
+    local hasGroup = allTypeRestrictions and rawget( allTypeRestrictions, group )
+
+    if hasGroup then
+        ULib.tsayError(ply, "Your rank is restricted from all ".. restrictionTypePlural)
```

Comment:
```suggestion
        ULib.tsayError( ply, "Your rank is restricted from all ".. restrictionTypePlural )
```

Reply:
Also TODO: Should probably localize `tsayError`

Reply:
Localized it